Source/ResponseSerialization.swift

```diff
@@ -144,25 +143,57 @@ extension DataRequest {
-                                                                       error: self.error) }
-                let end = CFAbsoluteTimeGetCurrent()
-
-                let response = DataResponse(request: self.request,
-                                            response: self.response,
-                                            data: self.data,
-                                            metrics: self.metrics,
-                                            serializationDuration: (end - start),
-                                            result: result)
-
-                self.eventMonitor?.request(self, didParseResponse: response)
-
-                (queue ?? .main).async { completionHandler(response) }
+        appendResponseSerializer {
+            let start = CFAbsoluteTimeGetCurrent()
+            let result = Result { try responseSerializer.serialize(request: self.request,
+                                                                   response: self.response,
+                                                                   data: self.data,
+                                                                   error: self.error) }
+            let end = CFAbsoluteTimeGetCurrent()
+
+            let queue: DispatchQueue = queue ?? .main
+            let response = DataResponse(request: self.request,
+                                        response: self.response,
+                                        data: self.data,
+                                        metrics: self.metrics,
+                                        serializationDuration: (end - start),
+                                        result: result)
+
+            self.eventMonitor?.request(self, didParseResponse: response)
+
+            if let serializerError = result.error, let delegate = self.delegate {
+                delegate.retryResult(for: self, dueTo: serializerError) { retryResult in
+                    var didComplete: (() -> Void)? = { completionHandler(response) }
+
+                    if let retryResult = retryResult {
+                        switch retryResult {
+                        case .doNotRetry:
+                            break
+
+                        case .doNotRetryWithError(let retryError):
+                            let result = Result<Serializer.SerializedObject>.failure(retryError)
+
+                            let response = DataResponse(request: self.request,
+                                                        response: self.response,
+                                                        data: self.data,
+                                                        metrics: self.metrics,
+                                                        serializationDuration: (end - start),
+                                                        result: result)
+
+                            didComplete = { completionHandler(response) }
+
+                        default:
```

Comment:
We should probably be explicit here since this `default` only handles a single extra case.

Reply:
Good call, switched in fea5fc86.